lib/mini_racer/truffleruby.rb

```diff
@@ -0,0 +1,306 @@
+# frozen_string_literal: true
+
+require "monitor"
+
+module MiniRacer
+
+  class Context
+    def heap_stats
+        {
+            total_physical_size: 0,
+            total_heap_size_executable: 0,
+            total_heap_size: 0,
+            used_heap_size: 0,
+            heap_size_limit: 0,
+
+        }
+    end
+
+    def stop
+        if @context.respond_to?(:stop)
+          if @entered
+            @context.stop
+            @stopped = true
+            stop_attached
+          end
+        end
+    end
+
+    private
+
+    @@context_initialized = false
+    @@use_strict = false
+
+    def init_unsafe(isolate, snapshot)
+        unless defined?(Polyglot::InnerContext)
+            raise "TruffleRuby #{RUBY_ENGINE_VERSION} does not have support for inner contexts, use a more recent version"
+        end
+    
+        unless Polyglot.languages.include? "js"
+            raise "The language 'js' is not available, you likely need to `export TRUFFLERUBYOPT='--jvm --polyglot'`" +
+                  "Note that you need TruffleRuby+GraalVM and not just the TruffleRuby standalone to use #{self.class}"
+        end
+
+        @context = Polyglot::InnerContext.new
+        @@context_initialized = true
+        @js_object = @context.eval('js', 'Object')
+        @isolate_mutex = Monitor.new
```

Comment:
A `Monitor` was used due to seeing errors:
```
MiniRacerTest#test_attached_exceptions:
ThreadError: deadlock; recursive locking
    /Users/bfish/Documents/projects/mini_racer-bjfish/lib/mini_racer.rb:222:in `synchronize'
    /Users/bfish/Documents/projects/mini_racer-bjfish/lib/mini_racer.rb:222:in `eval'
    /Users/bfish/Documents/projects/mini_racer-bjfish/lib/mini_racer/truffleruby.rb:123:in `notify_v8'
    /Users/bfish/Documents/projects/mini_racer-bjfish/lib/mini_racer.rb:144:in `initialize'
    /Users/bfish/Documents/projects/mini_racer-bjfish/lib/mini_racer.rb:296:in `block in attach'
    /Users/bfish/Documents/projects/mini_racer-bjfish/lib/mini_racer.rb:295:in `synchronize'
    /Users/bfish/Documents/projects/mini_racer-bjfish/lib/mini_racer.rb:295:in `attach'
    /Users/bfish/Documents/projects/mini_racer-bjfish/test/mini_racer_test.rb:184:in `test_attached_exceptions'
```

Reply:
Suspicious, we should see why it does not fail with a Mutex in the C ext

Reply:
@eregon I think their implementation of `notify_v8` doesn't call eval(which locks the mutex) like I implemented for `TruffleRuby`: https://github.com/rubyjs/mini_racer/blob/746bfddfeacf671bea7d7c7dd7897a7edcc452fd/ext/mini_racer_extension/mini_racer_extension.cc#L1334

Reply:
Check if still needed